internal/usage/handler.go

```diff
@@ -0,0 +1,162 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/crossplane/crossplane-runtime/pkg/controller"
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	xpunstructured "github.com/crossplane/crossplane-runtime/pkg/resource/unstructured"
+
+	"github.com/crossplane/crossplane/apis/apiextensions/v1alpha1"
+	"github.com/crossplane/crossplane/internal/features"
+)
+
+const (
+	// InUseIndexKey used to index CRDs by "Kind" and "group", to be used when
+	// indexing and retrieving needed CRDs
+	InUseIndexKey = "inuse.apiversion.kind.name"
+
+	// Error strings.
+	errFmtUnexpectedOp = "unexpected operation %q, expected \"DELETE\""
+)
+
+// IndexValueForObject returns the index value for the given object.
+func IndexValueForObject(u *unstructured.Unstructured) string {
+	return fmt.Sprintf("%s.%s.%s", u.GetAPIVersion(), u.GetKind(), u.GetName())
+}
+
+// SetupWebhookWithManager sets up the webhook with the manager.
+func SetupWebhookWithManager(mgr ctrl.Manager, options controller.Options) error {
+	if !options.Features.Enabled(features.EnableAlphaUsages) {
+		return nil
+	}
+	indexer := mgr.GetFieldIndexer()
+	if err := indexer.IndexField(context.Background(), &v1alpha1.Usage{}, InUseIndexKey, func(obj client.Object) []string {
+		u := obj.(*v1alpha1.Usage)
+		if u.Spec.Of.ResourceRef == nil || len(u.Spec.Of.ResourceRef.Name) == 0 {
+			return []string{}
+		}
+		return []string{fmt.Sprintf("%s.%s.%s", u.Spec.Of.APIVersion, u.Spec.Of.Kind, u.Spec.Of.ResourceRef.Name)}
+	}); err != nil {
+		return err
+	}
+
+	mgr.GetWebhookServer().Register("/validate-no-usages",
+		&webhook.Admission{Handler: NewHandler(
+			xpunstructured.NewClient(mgr.GetClient()),
+			WithLogger(options.Logger.WithValues("webhook", "no-usages")),
+		)})
+	return nil
+}
+
+// Handler implements the admission Handler for Composition.
+type Handler struct {
+	reader client.Reader
+	log    logging.Logger
+}
+
+// HandlerOption is used to configure the Handler.
+type HandlerOption func(*Handler)
+
+// WithLogger configures the logger for the Handler.
+func WithLogger(l logging.Logger) HandlerOption {
+	return func(h *Handler) {
+		h.log = l
+	}
+}
+
+// NewHandler returns a new Handler.
+func NewHandler(reader client.Reader, opts ...HandlerOption) *Handler {
+	h := &Handler{
+		reader: reader,
+		log:    logging.NewNopLogger(),
+	}
+
+	for _, opt := range opts {
+		opt(h)
+	}
+
+	return h
+}
+
+// Handle handles the admission request, validating there is no usage for the
+// resource being deleted.
+func (h *Handler) Handle(ctx context.Context, request admission.Request) admission.Response {
+	switch request.Operation {
+	case admissionv1.Create, admissionv1.Update, admissionv1.Connect:
```

Comment:
We are not handling them, we don't expect them since webhook configuration does not list them and error out as "this is unexpected" if it happens. Also see https://github.com/crossplane/crossplane/pull/4444#discussion_r1293745304